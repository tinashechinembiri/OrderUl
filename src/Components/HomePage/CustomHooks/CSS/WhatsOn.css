.input-bars{
 
    display: inline;
    padding: 5px;
    justify-content: space-between;
    white-space: nowrap;
    padding-left: 2px;
 
}

.location-input{
    position: relative;
    display:inline;
    margin-right:10px; 
    width: 8%;
    height: 2%;
    background-color: blue;
    padding-right: 14px;

}
.movie-input{
    display: inline;
    width: 8%;
    height: 2%;
    background-color: purple;
}
.search-background{
    padding: 2px;
    background-color: pink;
}
.display-header{
    display: inline-block;
    margin-right: 50px;
    width: 8%;
    height: 2%;
    font-family:'Courier New', Courier, monospace;
    size: 20px;
    
}
.advert-panel
{
    display: inline;
    left: 50%;
    background-color: maroon;
}
.searchbars-total{
    position: relative;
    left: 15%;
   
    display: inline;
    padding-bottom: 30px;
}