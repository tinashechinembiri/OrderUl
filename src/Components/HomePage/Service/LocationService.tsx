

const Dummy_Places :Array<{id:string, Name:string}> =  [
    {id : 'L1' , Name:'Leeds Cinema' }, 
    {id : 'N1' , Name:'Northampton Cinema' }, 
    {id : 'M1' , Name:'Manchester Cinema' }, 
    {id : 'C1' , Name:'Chester Cinema' }, 
    {id : 'LN1' , Name:'London Cinema' }, 
    {id : 'B1' , Name:'Birmingham Cinema' },     
]; 
const GetLocationdata = async () =>  {
     return await Dummy_Places ; 
}

export default GetLocationdata; 