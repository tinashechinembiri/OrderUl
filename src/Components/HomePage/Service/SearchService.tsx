import React from 'react';
const Dummy_Movies =[
    {id:1, Name: 'spaceranger', MovieDetails:{
        Cast :['Actor 1', 'actor 2'], 
        Synopsis:'movies is about items and all that ', 
        Director:['direct 1']
    }}, 
    {id:2, Name: 'powerranger', MovieDetails:{
        Cast :['Actor 2', 'actor 4'], 
        Synopsis:'movies is about items and all that ', 
        Director:['direct 2']
    }}, 
    {id:3, Name: 'promibe', MovieDetails:{
        Cast :['Actor 2', 'actor 4'], 
        Synopsis:'movies is about items and all that ', 
        Director:['direct 2']
    }}
]
const SearchService = async () => {
const movies =  Dummy_Movies; 

return movies; 
}
export default SearchService; 