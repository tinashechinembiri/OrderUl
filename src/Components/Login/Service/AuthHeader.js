export default authHeader =() => {
    // get the current user user const user = JSON.parse(localStorage.getItem('user'));
    // check if the user exist and return the token {'x-access-token': user.accessToken}
    // if failed return as empty object {}
    

}