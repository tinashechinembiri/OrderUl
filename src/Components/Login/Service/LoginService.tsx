import React from 'react';
import{actionType} from '../../../Helpers/Reducer'; 
import axios from 'axios';
import getRandomInt from './Helpers/Registerhelpers'


const Login_Service_Url  = 'http://localhost:1337/api/auth/'
// observer could help 
interface IResponse{
    status:number, 
    
}
export interface Icreateuser{
    email : string, 
    password : string, 
    firstname:string, 
    lastname:string, 
    phone:string, 
    addressname:string, 
    postcode:string, 
    county:string, 
    country:string
}
const LoginService = async (user : {username:string, password:string}, dispatch:React.Dispatch<actionType> ) => {
        const username = user.username; 
        const password =user.password; 
        const requestOptions = {
            method:'post',
           crossDomain:true,
            headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin':'*',
            'Access-Control-Allow-Methods':'POST,PATCH,OPTIONS',  },
            body : JSON.stringify({username, password})
        }
        try {
             const response =  await fetch( Login_Service_Url+'login/', 
             requestOptions); 
             const {code, token, message} = await response.json(); 
             if (code as number !== 200)
             {
                dispatch({type:'error', data: {_errors:message}}); 
                
                 return {error:message}    
             } 
             dispatch({type:'success', data: {user:{token:token, username:username}}})
             localStorage.setItem('user', JSON.stringify({token, username:username}));
        }
        catch(ex:any)
        {
         dispatch ({type:'error', data: {_errors:ex}}); 
        }      
}; 

const register = async(registerData:Icreateuser) => {
 const headers = { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin':'*',
    'Access-Control-Allow-Methods':'POST,PATCH,OPTIONS',  }
    const _userData = createregisterData(registerData); 
    axios.post(Login_Service_Url+'signup/',_userData, {headers})
    .then((response:any) => {
       if (response.code !== 200)
       {
            return false; 
       }
       return true 
    }).catch((ex:any)=> {
        console.log(ex); 

    })
    
}

const createregisterData = (registerData:Icreateuser) => {
    const accountid = 'ac'+getRandomInt(10000) as string; 
    const user_id = 'user_'+getRandomInt(10000) as string; 
    const _createdUser = {
        accountid, 
        user_id, 
        ...registerData
        // user_id: user_id, 
        // email:registerData.email, 
        // password:registerData.password, 
        // Account:{
        //     accountid: accountid, 
        //     firstname:registerData.firstname, 
        //     lastname:registerData.lastname, 
        //     phone:registerData.phone, 
        //     Address :{
        //         addressname:registerData.addressname, 
        //         postcode:registerData.postcode, 
        //         county:registerData.county, 
        //         country:registerData.country
        //     }
        // }  
    }
    return _createdUser; 
}

const logout = (dispatch:React.Dispatch<actionType>) => {
    try{
    dispatch({type:'logout'})
    // remove login from  local storage 
    localStorage.removeItem('user'); 
    }catch (ex :any)
    {
        console.log(ex); 
    }
    
}

const getCurrentUser = () => {
    // return json.parse from local storage 
    try{
        let value = localStorage.getItem('user'); 
        if (typeof value === 'string') {
            const parse = JSON.parse(value); 
            return parse; 
        }
    }catch (ex:any) 
    {
    
    } 
}
export  { LoginService , logout, getCurrentUser,register }; 