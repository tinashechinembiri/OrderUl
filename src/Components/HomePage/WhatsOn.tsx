import React, { useState, useEffect, lazy, Suspense } from "react";
import CinemaModal from './CinemaModal'
import useModal from "./CustomHooks/UseModal";
import './CustomHooks/CSS/WhatsOn.css'; 
import GetLocationdata from  './Service/LocationService'

const AdvertPanel = lazy(() => import('./WhatsOnPanel/AdvertPanel')); 
const Search = lazy(() => import('./Search/Search'));
const WhatsOn = () => {
    const [location, setLocation] = useState<string|null>(''); // using array destructuring 
    const {isShowing, toggle} =   useModal(); // using object destructuring 
    const [cinemaPlaces, setcinemaPlaces] = useState<Array<{id:string, Name:string}>>([]); 

  

    const setSelectedLocation = (location : string| null) => {
        setLocation(location);
        if(location)
        {
            toggle(); 
        }
    }
    const getlocation = async () => {
        const data = await GetLocationdata(); 
        setcinemaPlaces(data)
    }
    useEffect( () =>{
        getlocation();  
    }, []);
    return(
        <React.Fragment>
            <Suspense fallback={<div>Loading...</div>}>
                <div className="input-bars">
                    <div className="searchbars-total">
                        <div role ="heading" aria-level={2}className="display-header"> Whats on</div>  
                    <div className="movie-input">
                        <input className="location-input" onClick={toggle} readOnly value={!location  ? 'choose your location': location}/>                                  
                     <CinemaModal     
                         isShowing={isShowing}
                         hide={toggle}  
                         Dummy_Places_data={cinemaPlaces}  
                         setSelectedLocation={setSelectedLocation}
                         />    
                        <Search/>
                        </div>
                   </div>
                   
                         <AdvertPanel/>
                        
                </div> 
                </Suspense>        
        </React.Fragment>
    );
}

export default WhatsOn; 