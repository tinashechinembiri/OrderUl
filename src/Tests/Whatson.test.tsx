import React, {ReactNode ,useState  ,lazy, Suspense}from 'react';
import WhatsOn from '../Components/HomePage/WhatsOn';
import * as GetLocationdata from '../Components/HomePage/Service/LocationService'
import { render, screen, act, waitFor} from '@testing-library/react';
import ReactDOM from 'react-dom';
import {create} from 'react-test-renderer';
import userEvent from '@testing-library/user-event'
//jest.mock('../Components/HomePage/WhatsOnPanel/AdvertPanel', () => <AdvertPanel/>);
//jest.mock('../Components/HomePage/Search/Search', () => require('./Search'));
//https://github.com/testing-library/react-hooks-testing-library/issues/425



//https://github.com/testing-library/react-testing-library/issues/281
describe ('whats on' , () => {
    let locationdata ; 
    const oldCreatePortal = ReactDOM.createPortal;
    beforeAll(()=> {
        ReactDOM.createPortal = (node: ReactNode): React.ReactPortal =>
        node as React.ReactPortal;
    })
    afterAll(()=> {
        ReactDOM.createPortal = oldCreatePortal; 
       
    })
    beforeEach (() => {
        //locationdata = () => Promise.resolve([ {id : 'L1' , Name:'Leeds Cinema' }])
        const mock = jest.spyOn (GetLocationdata,'default').mockResolvedValue([
            {
                id : 'B1' , 
                Name:'Birmingham Cinema' 
            }]);
    })
    afterEach (() => {
       
        jest.clearAllMocks(); 

    })
  
    it ('load the whats on page', async () => {
    await act (async ()=> {
         render (
            <Suspense fallback={<div>loading...</div>}>
                <WhatsOn/> 
            </Suspense>);
              expect(screen.getByText('Loading...')).toBeInTheDocument();
          
            }); 
     //  expect( getByRole('heading')).toBeInTheDocument();
           await waitFor (() => {
            expect( screen.getByText('Comming soon')).toBeTruthy(); 
           })
        
      
    })
    
    it ('load the whats on page - test', async() => { 
        // const test = create (
        //     <Suspense fallback={<div>loading...</div>}>

        //         <WhatsOn/> 
        //         </Suspense>
               
        // )
       
       // expect(test.toJSON()).toMatchSnapshot(); 
    }); 
    //https://www.carlrippon.com/how-to-mock-a-function-in-jest-with-typescript/
    //https://davidwcai.medium.com/react-testing-library-and-the-not-wrapped-in-act-errors-491a5629193b
 
    it ('user should be able to interact with the location modal', async() => {
    
             await waitFor (() => {
                  render ( 
                 <Suspense fallback={<div>loading...</div>}>
                 <WhatsOn/>
                 </Suspense>
                 ) 
             }); 
             await waitFor (() => {
                userEvent.click(screen.getByDisplayValue('choose your location'))
             })         
    }) 

})