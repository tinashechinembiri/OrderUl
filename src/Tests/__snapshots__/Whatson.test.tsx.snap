// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whats on load the whats on page - test 1`] = `
<div
  className="input-bars"
>
  <div
    className="searchbars-total"
  >
    <div
      aria-level={2}
      className="display-header"
      role="heading"
    >
       Whats on
    </div>
    <div
      className="movie-input"
    >
      <input
        className="location-input"
        onClick={[Function]}
        readOnly={true}
        value="choose your location"
      />
      <input
        className="search_input"
        id="movie-search"
        onChange={[Function]}
        placeholder="Start typing a film or event"
        type="text"
        value=""
      />
    </div>
  </div>
  <ul
    className="ul-list-adverts"
  />
</div>
`;
